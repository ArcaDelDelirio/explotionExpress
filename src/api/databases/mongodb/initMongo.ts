import { connect } from 'mongoose';

export class SingleMongo {
    private readonly dbURI: string;
    public constructor(url: string, collection: string) {
        this.dbURI = `mongodb://${url}/${collection}`;
    }

    public connectToMongo() {
        return connect(this.dbURI);
    }
}
