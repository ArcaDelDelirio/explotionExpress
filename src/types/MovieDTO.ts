export default interface MovieDTO {
    title: string;
    year: number;
    genre: string[];
    duration: number;
}
