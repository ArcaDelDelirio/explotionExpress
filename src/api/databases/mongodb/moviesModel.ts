import { Document, Model, model, Schema } from 'mongoose';
import { MovieDTO } from 'MovieDTO';

export interface MovieModel extends MovieDTO, Document {
    getTitleAndYear(): string;
}
export const MovieSchema: Schema = new Schema({
    title: String,
    year: Number,
    genre: [String],
    duration: Number,
});

MovieSchema.methods.getTitleAndYear = function(): string {
    return this.title + ' ' + this.duration;
};

export const Movie: Model<MovieModel> = model<MovieModel>('movies', MovieSchema);
