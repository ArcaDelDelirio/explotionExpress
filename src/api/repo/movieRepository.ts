import { Movie, MovieModel } from '../databases/mongodb/moviesModel';
import { SingleMongo } from '../databases/mongodb/initMongo';
import { MongoRepository } from './MongoRepository';
import MovieDTO from 'MovieDTO';
import { ObjectId } from 'bson';

export class MovieRepository implements MongoRepository<MovieModel,MovieDTO> {

    //private data: MovieDTO;
    private state: number;
    private con: SingleMongo;
    private model = Movie;

    public constructor(databaseURI: string, databaseName: string) {
        this.con = new SingleMongo(databaseURI, databaseName);
        this.con.connection.on('error', console.error.bind(console, 'connection error:'));
        this.con.connection.once('open', () => console.log('We are connected'));
    }
    public async getConnection() {
        return await this.con.connection;
    }

    public async getAll() {
        return await this.model.find({}).limit(10);
    }

    public async create(data: MovieDTO){
        const record = new Movie(data)
        return await record.save()
    }

    public async getById(id: ObjectId | number){
        return await this.model.findById(id)
    }

    public async updateById(id: ObjectId | number,data: {}){
        return await this.model.findByIdAndUpdate(id,data)
    }

    public async deleteById(id: ObjectId | number){
        return await this.model.findByIdAndDelete(id)
    }

    public async giveLikeById(id: ObjectId | number ){
        const likes = 1
        return await this.model.findByIdAndUpdate(id,likes)
    }

    public async getByTitle(title: string){
        return await this.model.findById(title)
    }

    public async updateByTitle(title: string,data: {}){
        return await this.model.findByIdAndUpdate(title,data)
    }
    //TODO: Implements title - years - and all Model and Interfaces Types
}
