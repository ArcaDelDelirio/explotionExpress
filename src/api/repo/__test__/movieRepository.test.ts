import { MovieRepository } from '../movieRepository';
import MovieDTO from 'MovieDTO';
import { ObjectId } from 'mongodb';

describe('MovieRepository', () => {
    //TODO: This can be mock
    const repo = new MovieRepository('127.0.0.1','movies');
    const newMovie: MovieDTO = {
        title:'Interstellar',
        year:2014,
        genre:['science','fiction'],
        duration: 169
    }
    describe('getConnectionState', () => {
        test('Check connection to database', () => {
            return repo.getConnection().then(state => {
                expect(state.readyState).toBe(1);
            });
        });
    });
    describe('createMovie', () => {
        test('should return void or callback', () => {
            return repo.create(newMovie)
            .then(data => {
                console.log("Create :",data)
                expect(data).toMatchSnapshot();
            });
        });
    });
    describe('getMovieById', () => {
        test('should return data for Id', () => {
            const id = new ObjectId('5d305371610ea402dee6547b')
            return repo.getById(id)
            .then(data => {
                console.log("Get :",data)
                expect(data).toMatchSnapshot();
            });
        });
    });
    describe('updateMovieById', () => {
        test('should return json object updated', () => {
            const id = new ObjectId('5d305371610ea402dee6547b')
            return repo.updateById(id,{title:'Georgia'})
            .then(data => {
                console.log("Update: ",data)
                expect(data).toMatchSnapshot();
            });
        });
    });
    describe('deleteMovieById', () => {
        test('should return a delete response with obj and number', () => {
            const id = new ObjectId('5d305371610ea402dee6547b')
            return repo.deleteById(id)
            .then(data => {
                console.log("Delete: ",data)
                expect(data).toBe(data);
            });
        });
    });
    describe('giveLikeId', () => {
        test('should return a new property for object', () => {
            const id = new ObjectId('5d305371610ea402dee6547d')
            return repo.giveLikeById(id)
            .then(data => {
                console.log("Give: ",data)
                expect(data).toBe(data);
            });
        });
    });
    describe('getAllMovies', () => {
        test('should return all movies in JSON', () => {
            return repo.getAll()
            .then(data => {
                console.log("All: ",data)
                expect(data).toMatchSnapshot();
            });
        });
    });
});
