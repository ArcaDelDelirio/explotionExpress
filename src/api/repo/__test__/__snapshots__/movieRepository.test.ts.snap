// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieRepository createMovie should return void or callback 1`] = `
Object {
  "__v": 0,
  "_id": "5d3054fcf7b5af25f4eacae8",
  "duration": 169,
  "genre": Array [
    "science",
    "fiction",
  ],
  "title": "Interstellar",
  "year": 2014,
}
`;

exports[`MovieRepository getAllMovies should return all movies in JSON 1`] = `
Array [
  Object {
    "_id": "5d305371610ea402dee6547c",
    "duration": 181,
    "genre": Array [
      "SuperHero",
      "Action",
    ],
    "title": "Vengadores:Endgame",
    "year": 2019,
  },
  Object {
    "_id": "5d305371610ea402dee6547d",
    "duration": 129,
    "genre": Array [
      "Terror",
      "Suspense",
    ],
    "title": "Glass",
    "year": 2019,
  },
  Object {
    "__v": 0,
    "_id": "5d305377066cd610945a334f",
    "duration": 169,
    "genre": Array [
      "science",
      "fiction",
    ],
    "title": "Interstellar",
    "year": 2014,
  },
  Object {
    "__v": 0,
    "_id": "5d3053e75bb36122d8c2dbf1",
    "duration": 169,
    "genre": Array [
      "science",
      "fiction",
    ],
    "title": "Interstellar",
    "year": 2014,
  },
  Object {
    "__v": 0,
    "_id": "5d3053ea11c0aa22d8d7ee60",
    "duration": 169,
    "genre": Array [
      "science",
      "fiction",
    ],
    "title": "Interstellar",
    "year": 2014,
  },
  Object {
    "__v": 0,
    "_id": "5d3054fcf7b5af25f4eacae8",
    "duration": 169,
    "genre": Array [
      "science",
      "fiction",
    ],
    "title": "Interstellar",
    "year": 2014,
  },
]
`;

exports[`MovieRepository getMovieById should return data for Id 1`] = `
Object {
  "_id": "5d305371610ea402dee6547b",
  "duration": 100,
  "genre": Array [
    "Comedy",
    "Animation",
  ],
  "title": "Toy Story 4",
  "year": 2019,
}
`;

exports[`MovieRepository updateMovieById should return json object updated 1`] = `
Object {
  "_id": "5d305371610ea402dee6547b",
  "duration": 100,
  "genre": Array [
    "Comedy",
    "Animation",
  ],
  "title": "Toy Story 4",
  "year": 2019,
}
`;
